### PR TITLE
Implement SOCKS proxy functionality

### PR DESCRIPTION
**Pull request by Davidde94**
Add a new Proxy type to enable a `HTTPClient` to send requests via a SOCKSv5 Proxy server.

### CODE REVIEW

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift [hunk 1]
```diff
@@ -50,7 +59,11 @@ public extension HTTPClient.Configuration {
         ///     - port: proxy server port.
         ///     - authorization: proxy server authorization.
         public static func server(host: String, port: Int, authorization: HTTPClient.Authorization? = nil) -> Proxy {
-            return .init(host: host, port: port, authorization: authorization)
+            return .init(host: host, port: port, authorization: authorization, type: .http)
+        }
+        
+        public static func socksServer(host: String, port: Int = 1080, httpAuthorization: HTTPClient.Authorization? = nil) -> Proxy {
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:65]: Presumably we don't want the `httpAuthorization` field.

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 2]
```diff
@@ -899,6 +900,13 @@ extension ChannelPipeline {
         try sync.addHandler(decoder)
         try sync.addHandler(handler)
     }
+    
+    func syncAddSOCKSProxyHandler() throws {
+        let address = try SocketAddress(ipAddress: "127.0.0.1", port: 12345)
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClient.swift:905]: This hardcoded address seems bogus.

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+@testable import AsyncHTTPClient
+import NIO
+import NIOHTTP1
+import NIOSOCKS
+import XCTest
+
+class MockSocksServer {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:21]:
```suggestion
class MockSOCKSServer {
```

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 4]
```diff
@@ -899,6 +900,13 @@ extension ChannelPipeline {
         try sync.addHandler(decoder)
         try sync.addHandler(handler)
     }
+    
+    func syncAddSOCKSProxyHandler(host: String, port: Int) throws {
+        let address = try SocketAddress(ipAddress: host, port: port)
+        let handler = SOCKSClientHandler(targetAddress: .address(address))
+        let sync = self.syncOperations
+        try sync.addHandler(handler)
+    }
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/HTTPClient.swift:909]: Is this something for NIO extras?

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 5]
```diff
@@ -708,6 +709,23 @@ class HTTPClientTests: XCTestCase {
...
+        let localClient = HTTPClient(eventLoopGroupProvider: .shared(self.clientGroup),
+                                     configuration: .init(proxy: .socksServer(host: "127.0.0.1")))
+        
+        do  {
+            let response = try localClient.get(url: "http://127.0.0.1/socks/test").wait()
+            XCTAssertEqual(.ok, response.status)
+            XCTAssertEqual(ByteBuffer(string: "it works!"), response.body)
+        } catch {
+            XCTFail("\(error)")
+        }
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:724]: Super NIT. Johannes showed me this pattern some time ago. Personally I prefer it.

```suggestion
        var response: HTTPClient.Response?
        XCTAssertNoThrow(response = try localClient.get(url: "http://127.0.0.1/socks/test").wait())
        XCTAssertEqual(.ok, response?.status)
        XCTAssertEqual(ByteBuffer(string: "it works!"), response?.body)
```

File: Sources/AsyncHTTPClient/HTTPClient.swift (hunk 4 above)
Comment (Lukasa) [reply Sources/AsyncHTTPClient/HTTPClient.swift:909]: We don't normally offer those extensions for protocols comprised of a single handler though.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift [hunk 6]
```diff
@@ -14,6 +14,7 @@
 
 import NIO
 import NIOHTTP1
+import NIOSOCKS
 
 public extension HTTPClient.Configuration {
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:19]: Can we delete the word `public` from here and move it to `struct Proxy`? It's making it hard to reason about the visibility of things.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift [hunk 7]
```diff
@@ -27,20 +28,28 @@ public extension HTTPClient.Configuration {
...
+            case http
+            case socks
+        }
+        
         /// Specifies Proxy server host.
         public var host: String
         /// Specifies Proxy server port.
         public var port: Int
         /// Specifies Proxy server authorization.
         public var authorization: HTTPClient.Authorization?
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:42]: I think authorisation should be a property of the `ProxyType`. We have different auth requirements for HTTP and SOCKS proxies.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift [hunk 8]
```diff
@@ -50,7 +59,11 @@ public extension HTTPClient.Configuration {
         ///     - port: proxy server port.
         ///     - authorization: proxy server authorization.
         public static func server(host: String, port: Int, authorization: HTTPClient.Authorization? = nil) -> Proxy {
-            return .init(host: host, port: port, authorization: authorization)
+            return .init(host: host, port: port, authorization: authorization, type: .http)
+        }
+        
+        public static func socksServer(host: String, port: Int = 1080) -> Proxy {
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:65]: Can you add a doc comment here?

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 9]
```diff
@@ -708,6 +709,23 @@ class HTTPClientTests: XCTestCase {
             }
         }
     }
+    
+    func testProxySOCKS() throws {
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:713]: Can we have some more tests here? Can we validate that attempting to speak SOCKS to a server that doesn't speak SOCKS fails? Can we validate that we handle various handshake failure modes appropriately?

File: Sources/AsyncHTTPClient/HTTPClient.swift (hunk 4 above)
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClient.swift:909]: This is also slightly different to the pattern I think you mean, here I've just replicated what is done for the HTTP proxy above.

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@testable import AsyncHTTPClient
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:15]: Why does this need to be a testable import?

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 11]
```diff
@@ -0,0 +1,111 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2020 Apple Inc. and the AsyncHTTPClient project authors
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:5]:
```suggestion
// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
```

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+class MockSocksServer {
+    
+    let channel: Channel
+    
+    public init(expectedURL: String, expectedResponse: String, file: String = (#file), line: UInt = #line) throws {
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        let bootstrap = ServerBootstrap.init(group: elg)
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelInitializer { channel in
+            let handshakeHandler = SOCKSServerHandshakeHandler()
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:30]: Can you clean up the indentation here?

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 13]
```diff
@@ -708,6 +709,23 @@ class HTTPClientTests: XCTestCase {
...
+        do  {
+            let response = try localClient.get(url: "http://127.0.0.1/socks/test").wait()
+            XCTAssertEqual(.ok, response.status)
+            XCTAssertEqual(ByteBuffer(string: "it works!"), response.body)
+        } catch {
+            XCTFail("\(error)")
+        }
+        
+        XCTAssertNoThrow(try localClient.syncShutdown())
+        XCTAssertNoThrow(try socksBin.shutdown())
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:727]: These two should probably be in `defer` blocks.

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        self.expectedResponse = expectedResponse
+        self.file = file
+        self.line = line
+    }
+    
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let message = self.unwrapInboundIn(data)
+        switch message {
+        case .head(let head):
+            XCTAssertEqual(head.uri, self.expectedURL)
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:67]: Should this be defensive against receiving only one request?

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    
+    init(handshakeHandler: SOCKSServerHandshakeHandler) {
+        self.handshakeHandler = handshakeHandler
+    }
+    
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let message = self.unwrapInboundIn(data)
+        switch message {
+        case .greeting:
+            context.writeAndFlush(.init(
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:92]: Presumably this flush isn't needed.

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+                ServerMessage.selectedAuthenticationMethod(.init(method: .noneRequired))), promise: nil)
+            context.writeAndFlush(.init(
+                ServerMessage.authenticationData(context.channel.allocator.buffer(capacity: 0), complete: true)), promise: nil)
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: Do we need to send authentication data here?

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        case .authenticationData:
+            break
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:97]: This should presumably be an error.

**Review by Lukasa — CHANGES_REQUESTED** [review]
Cool, looking good! Some notes in the diff.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 6 above)
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:19]: Not really, I didn't add this, so it will break the public API.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 7 above)
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:42]: Again I can't really change this without breaking the public API :(

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:19]: No it won't. The only thing the `public` applies to here is the `struct`.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 7 above)
Comment (Lukasa) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:42]: Yes you can, but sorry, I wasn't clear.

I mean that `ProxyType` should only store the authorisation when the case is `http`. Then this property can become computed, and crash if it is set when the proxy type is set to `socks`.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 6 above)
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:19]: Ah, understood.

File: Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift (hunk 7 above)
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClientProxyHandler.swift:42]: Understood, thanks!

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift (hunk 16 above)
Comment (Davidde94) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: We need to send an empty buffer with the `complete` flag set to true so that the server state machine is updated.

Comment (Lukasa) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: Oh. Huh. I think the server should have a fast-path for `noneRequired` that can _tolerate_ receiving an empty buffer with `complete`, but will also happily progress without it.

Comment (Davidde94) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: Sure, I'll file an issue in swift-nio-extras, or do you think we should provide this fast path *now*?

Comment (Lukasa) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: I think we should provide it before we ship swift-nio-extras.

Comment (Davidde94) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: Alright

Comment (Davidde94) [reply Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:95]: https://github.com/apple/swift-nio-extras/pull/134

**Comment by Davidde94** [comment]
@swift-server-bot test this please

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 18]
```diff
@@ -900,6 +901,13 @@ extension ChannelPipeline {
         try sync.addHandler(handler)
     }
 
+    func syncAddSOCKSProxyHandler(host: String, port: Int) throws {
+        let address = try SocketAddress(ipAddress: host, port: port)
```
[the remark this answers is not in the document]
Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClient.swift:905]: Yeh you're right

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+import XCTest
+
+struct MockSOCKSError: Error, Hashable {
+    var description: String
+}
+
+class MockSOCKSServer {
+    let channel: Channel
+
+    public init(expectedURL: String, expectedResponse: String, misbehave: Bool = false, file: String = #file, line: UInt = #line) throws {
```
Comment (PeterAdams-A) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:28]: Why public?

**Review by PeterAdams-A — APPROVED** [review]
A few minor comments but looks generally OK to me.

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 20]
```diff
@@ -709,6 +710,69 @@ class HTTPClientTests: XCTestCase {
...
+    func testProxySOCKS() throws {
+        let socksBin = try MockSOCKSServer(expectedURL: "/socks/test", expectedResponse: "it works!")
+        let localClient = HTTPClient(eventLoopGroupProvider: .shared(self.clientGroup),
+                                     configuration: .init(proxy: .socksServer(host: "127.0.0.1")))
+
+        defer {
+            XCTAssertNoThrow(try localClient.syncShutdown())
+            XCTAssertNoThrow(try socksBin.shutdown())
+        }
+
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:722]: Can we move those tests into an extra test file `HTTPClient+SOCKSTests` or something along the lines? `HTTPClientTests` is slowly approaching 3k loc. And I think more separation will make dealing with test easier going forward.

File: Sources/AsyncHTTPClient/HTTPHandler.swift [hunk 21]
```diff
@@ -342,8 +342,8 @@ extension HTTPClient {
     }
 
     /// HTTP authentication
-    public struct Authorization {
-        private enum Scheme {
+    public struct Authorization: Hashable {
+        private enum Scheme: Hashable {
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/HTTPHandler.swift:346]: Why was this added?

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+            }
+        self.channel = try bootstrap.bind(host: "127.0.0.1", port: 1080).wait()
+    }
+
+    func shutdown() throws {
+        try self.channel.close().wait()
+    }
+}
+
+class SOCKSTestHTTPClient: ChannelInboundHandler {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:48]: I'm not sure about the name `SOCKSTestHTTPClient`. This handler receives `HTTPServerRequestPart` and responds with `HTTPServerResponsePart`. This looks like a plain `HTTPEchoHandler` to me. No SOCKS needed. And IMHO not a client at all :)

File: Sources/AsyncHTTPClient/HTTPHandler.swift (hunk 21 above)
Comment (Lukasa) [reply Sources/AsyncHTTPClient/HTTPHandler.swift:346]: Getting equatable conformance on the internal Proxy Type enum.

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 23]
```diff
@@ -900,6 +901,13 @@ extension ChannelPipeline {
         try sync.addHandler(handler)
     }
 
+    func syncAddSOCKSProxyHandler(host: String, port: Int) throws {
+        let address = try SocketAddress.makeAddressResolvingHost(host, port: port)
+        let handler = SOCKSClientHandler(targetAddress: .address(address))
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClient.swift:906]: I think you resolved my feedback here but this is still valid: I think we should be using the hostname form here.

Comment (Davidde94) [reply Sources/AsyncHTTPClient/HTTPClient.swift:906]: Oh I misunderstood - I switched to using `let address = try SocketAddress.makeAddressResolvingHost(host, port: port)`

File: Sources/AsyncHTTPClient/HTTPClient.swift [hunk 24]
```diff
@@ -900,6 +901,13 @@ extension ChannelPipeline {
         try sync.addHandler(handler)
     }
 
+    func syncAddSOCKSProxyHandler(host: String, port: Int) throws {
+//        let address = try SocketAddress.makeAddressResolvingHost(host, port: port)
```
Comment (Lukasa) [inline Sources/AsyncHTTPClient/HTTPClient.swift:905]: Why is this commented out?

File: Tests/AsyncHTTPClientTests/HTTPClient+SOCKSTests.swift [hunk 25]
```diff
@@ -0,0 +1,139 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/HTTPClient+SOCKSTests.swift:5]:
```suggestion
// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
```

File: Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelInitializer { channel in
+                let handshakeHandler = SOCKSServerHandshakeHandler()
+                return channel.pipeline.addHandlers([
+                    handshakeHandler,
+                    SOCKSTestHandler(handshakeHandler: handshakeHandler, misbehave: misbehave),
+                    TestHTTPServer(expectedURL: expectedURL, expectedResponse: expectedResponse, file: file, line: line),
+                ])
+            }
+        self.channel = try bootstrap.bind(host: "127.0.0.1", port: 1080).wait()
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/SOCKSTestUtils.swift:40]: Can we use `localhost` for this? You can get IPv6-only machines, and there's no reason to fail on them.

File: Tests/AsyncHTTPClientTests/HTTPClient+SOCKSTests.swift [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+        self.serverGroup = nil
+
+        XCTAssertNotNil(self.backgroundLogStore)
+        self.backgroundLogStore = nil
+    }
+ 
+    func testProxySOCKS() throws {
+        let socksBin = try MockSOCKSServer(expectedURL: "/socks/test", expectedResponse: "it works!")
+        let localClient = HTTPClient(eventLoopGroupProvider: .shared(self.clientGroup),
+                                     configuration: .init(proxy: .socksServer(host: "127.0.0.1")))
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/HTTPClient+SOCKSTests.swift:79]: Related to the `localhost` note below, it'd be good to use `localhost` for the SOCKS server hostname throughout this file too.

**Review by Lukasa — CHANGES_REQUESTED** [review]
Cool, a few cleanups needed here.

**Review by Lukasa — APPROVED** [review]
Ok, this LGTM. Let's get a swift-nio-extras release out the door and we can move forward here.

**Comment by Davidde94** [comment]
@swift-server-bot test this please

File: Package.swift [hunk 28]
```diff
@@ -21,22 +21,22 @@ let package = Package(
         .library(name: "AsyncHTTPClient", targets: ["AsyncHTTPClient"]),
     ],
     dependencies: [
-        .package(url: "https://github.com/apple/swift-nio.git", from: "2.27.0"),
+        .package(url: "https://github.com/apple/swift-nio.git", from: "2.29.0"),
         .package(url: "https://github.com/apple/swift-nio-ssl.git", from: "2.13.0"),
-        .package(url: "https://github.com/apple/swift-nio-extras.git", from: "1.3.0"),
+        .package(url: "https://github.com/apple/swift-nio-extras.git", from: "1.9.0"),
```
Comment (Lukasa) [inline Package.swift:26]:
```suggestion
        .package(url: "https://github.com/apple/swift-nio-extras.git", from: "1.9.1"),
```

Comment (Davidde94) [reply Package.swift]: Thanks, I actually just pushed a commit doing this too :)

File: Tests/AsyncHTTPClientTests/HTTPClientTests.swift [hunk 29]
```diff
@@ -22,6 +22,7 @@ import NIOConcurrencyHelpers
 import NIOFoundationCompat
 import NIOHTTP1
 import NIOHTTPCompression
+import NIOSOCKS
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTests.swift:25]: I don't think we need this here :)

**Review by fabianfett — APPROVED** [review]
Thanks! Looks awesome!
